### PR TITLE
RFD 3: Extended approval workflows

### PR DESCRIPTION
**Pull request by klizhentas**

### CODE REVIEW

File: rfd/0003-extended-approval-workflows.md
Comment (klizhentas) [reply rfd/0003-extended-approval-workflows.md:136]: This could be a UX improvement on the client side, sure, but having explicit flow would be a first step.

File: rfd/0003-extended-approval-workflows.md
Comment (klizhentas) [reply rfd/0003-extended-approval-workflows.md:193]: The request/response flow is already async, what's missing is the UX on the UI and CLI. We can add `--nowait` flag to return access_request ID and let users poll it manually. I will update this RFD with proposal:

```
$ tsh login --request-roles=dictator --nowait

Requested roles dictator, check the status by calling

$ tsh login --request-id=request-12345
Pending
$ tsh login --request-id=request-12345
Granted by Alice... logging in
```

Comment (klizhentas) [reply rfd/0003-extended-approval-workflows.md:193]: @webvictim how does this look? see above

File: rfd/0003-extended-approval-workflows.md
Comment (fspmarshall) [inline rfd/0003-extended-approval-workflows.md:271]: I don't see the `always_request_access` option mentioned anywhere else.  Could you elaborate on what this is?

File: rfd/0003-extended-approval-workflows.md
Comment (klizhentas) [inline rfd/0003-extended-approval-workflows.md:101]: Add cluster labels support here

**Review by fspmarshall — COMMENTED** [review]
Automatic approval is mentioned in multiple section titles, but I don't actually see any proposed changes to teleport that would cause a request to be approved automatically.  Is "automatic approval" just referring to the fact that plugins can be written to perform automatic approvals, or are we looking to support some kind of automatic approvals in teleport itself?  Ex:

```
$ tctl request auto-approve --where-role=contractor --grant-role=customer-a --ttl=1h
Users with role 'contractor' will be automatically granted role 'customer-a' for 1h...
Do you want to continue?  (y/n):
```

File: rfd/0003-extended-approval-workflows.md
Comment (fspmarshall) [reply rfd/0003-extended-approval-workflows.md:136]: I think that if we don't lazily generate the exact role, then we should limit the roles that can be applied to those that the user would have been able to request explicitly (e.g. in the case of `roles: ['^customer-.*$']`, the admin can only select from the set of roles that would match that pattern to fulfill the request).

I'm worried about human error generally, but I'm especially worried about decision fatigue during a crisis.  When things go wrong is not the time for admins to be trying to reason about which tags will and won't let their contractors escalate beyond their maximum allowable permissions (for example).

There should be a worst-case upper-limit to the amount of access a user can gain via the access request system.

Comment (klizhentas) [reply rfd/0003-extended-approval-workflows.md:136]: I would like to keep request part explicit, but add this as a convenience flag to `tsh ssh`

```bash
$ tsh ssh --request root@mongodb 
Pending access request
Granted
$ 
```

File: rfd/0003-extended-approval-workflows.md
Comment (klizhentas) [reply rfd/0003-extended-approval-workflows.md:271]: I forgot to remove, `always_request_access` please ignore it.

Comment (klizhentas) [reply rfd/0003-extended-approval-workflows.md:271]: By auto approval I mean that plugin can automatically approve the request based on some criteria, I will clarify that.

File: rfd/0003-extended-approval-workflows.md
Comment (webvictim) [reply rfd/0003-extended-approval-workflows.md:193]: Yeah, I think that's probably better - the waiting on the CLI or UI end isn't inherently bad, as anyone can start a terminal and leave it sat there waiting for approval, or open a browser tab and leave it in the background doing the same thing. It's just a question of them being notified when the requested has been actioned.

**Comment by webvictim** [comment]
Maybe we could also implement https://github.com/gravitational/teleport/issues/4309 as part of these improvements? The ability to limit the scope of _who_ can approve a request is pretty valuable, as it's currently a binary permission.

File: rfd/0003-extended-approval-workflows.md
Comment (awly) [inline rfd/0003-extended-approval-workflows.md:173]: will the list contain all of the roles?
what if there's thousands of them (auto-generated), or a user doesn't have permission to read all roles?

File: rfd/0003-extended-approval-workflows.md
Comment (benarent) [inline rfd/0003-extended-approval-workflows.md:381]:
```suggestion
based on the ticket number, they will see the clusters to access.

## Audit Log
All request events will be tracked by Teleports Audit Log. These will recorded under
access_request.  These are tracked using these two event codes. 

`T5000I` - AccessRequestCreateCode is the the access request creation code.
`T5001I` - AccessRequestUpdateCode is the access request state update code.

AccessRequestUpdateCode provides an option for plugins to update the state off a request
with extra meta data about its reason for who approved, or why it was denied. See (Teleport Plugin)[https://github.com/gravitational/teleport-plugins/blob/22334ec352bc62fc6bddd98f2284228442da73fb/access/access.go#L128] as an example. 
```

File: rfd/0003-extended-approval-workflows.md
Comment (benarent) [inline rfd/0003-extended-approval-workflows.md:153]:
```suggestion
#### Request access traits

For external systems to make better decisions on incoming request. The request should send forward 
on any claims/traits provided by the IdP. This is configurable by the Teleport Admin as part of Teleport Setup.
https://gravitational.com/teleport/docs/enterprise/sso/ssh-sso/.
```

File: rfd/0003-extended-approval-workflows.md
Comment (benarent) [reply rfd/0003-extended-approval-workflows.md:380]: Removed

File: rfd/0003-extended-approval-workflows.md
Comment (benarent) [reply rfd/0003-extended-approval-workflows.md:55]: Updated

File: rfd/0003-extended-approval-workflows.md
Comment (benarent) [reply rfd/0003-extended-approval-workflows.md:15]: updated

File: rfd/0003-extended-approval-workflows.md
Comment (benarent) [reply rfd/0003-extended-approval-workflows.md:283]: Correct: I've updated this.

**Comment by benarent** [comment]
I've updated the RFD with latest comments and developments during 5.0 release.  I'm going to merge the RFD and create open issues for parts of the RFD that are still open for development.

**Review by benarent — APPROVED** [review]

**Pull request merged**
